### PR TITLE
training hpo experiments

### PR DESCRIPTION
**Pull request by woodybury**
## Summary
- **ESR**: `esr(..., eps)` with `mean(y²)+ε`; **`ESRLoss`** for `custom_losses` (validation still uses classic `esr` with `eps=0`).
- **EMA**: optional via `learning.ema` in **`nam-full`** (Lightning `EMAWeightAveraging`).
- **Optimizer**: JSON `{"class", "kwargs"}` (e.g. AdamW) or flat dict → Adam.
- **Data**: optional train **`jitter`** (random window offset up to `ny//4`).
- **Tests** for the above.

## Config (no code changes required)
- **`trainer.gradient_clip_val`** in learning JSON → Lightning `Trainer`.
- **`lr_scheduler`** (e.g. **ReduceLROnPlateau** + `monitor: val_loss`) in model JSON → existing `configure_optimizers` path.

## Notes
I actually found the ESR as aux loss doesn't help but keeping it here we are curious to sweep it. It may be worth trying it as main loss instead. As val loss it does seem to help...

IDK that reduce on plateau makes any sence w/ 350 epochs. maybe `CosineAnnealingWarmRestarts` instead or what we had.

Not sure which of the others are helping, need to do some ablation and sweeps.

## Tried and Failed

tried a warmup wrapper around lr scheduler - seemed to just hurt

## Configs

data.json
```
{
   ....
   "train": {
       ...
       "jitter": true
    }
    ....
}
```

learning.json

```
{
  "train_dataloader": {
    "batch_size": 16,
    "shuffle": true,
    "pin_memory": false,
    "drop_last": true,
    "num_workers": 0
  },
  "val_dataloader": {
    "pin_memory": false
  },
  "trainer": {
    "accelerator": "auto",
    "devices": 1,
    "max_epochs": 350,
    "gradient_clip_val": 1.0
  },
  "trainer_fit_kwargs": {},
  "ema": {
    "enabled": true,
    "decay": 0.999,
    "use_buffers": true
  }
}
```

model.json
```
{
    "net": {
        "name": "WaveNet",
        "config": {
            ...legolas config
        }
    },
    "loss": {
        "val_loss": "esr",
        "mrstft_weight": 0.0005,
        "custom_losses": {
            "ESR": {
                "name": "nam.models.losses.ESRLoss",
                "kwargs": {
                    "eps": 1e-6
                },
                "weight": 0 // zeroed this out and it did better
            }
        }
    },
    "optimizer": {
        "class": "AdamW",
        "kwargs": {
            "lr": 0.004,
            "weight_decay": 3.17e-07
        }
    },
    "lr_scheduler": {
        "class": "ReduceLROnPlateau",
        "kwargs": {
            "factor": 0.6,
            "patience": 10,
            "cooldown": 5,
            "min_lr": 1e-6,
            "threshold": 1e-4,
            "threshold_mode": "rel"
        },
        "monitor": "val_loss"
    }
}
```

### CODE REVIEW

**Comment by woodybury** [comment]
TODO tests:
- [x] val_loss / monitor esr on/off - this one for sure helped frg-4
- [x] EMA on/off
- [x] AdamW vs Adam
- [x] jitter on/off. if helpful maybe sweep window offset
- [x] "gradient_clip_val": 1.0 on/off. sweep val?
- [x] lr_scheduler: expo vs. cosine. sweep rate?

**Comment by woodybury** [comment]
Closing this PR based on the findings below:
- [x] Jitter: not sig improvement
- [x] EMA + ADAMW: slight sig improvement but not worth added complexity - would require testing at different max_epochs
- [x] ESR loss: not helpful

What was helpful is already supported in the current configs - no code changes needed:
- [x] higher init LR (possibly 0.006)
- [x] OneCycleLR or CosineAnnealingLR schedulers
- [x] gradient_clip_val = 1 (didn't hurt and can be used in parallel trainings)
- [x] batch sizes 16, 8 or 4 (didn't hurt when LR adjusted accordingly and can be used in parallel trainings)

esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
esr_by_tone_models
